pkg/streamingpromql/operators/selectors/extend_range_vector.go

```diff
@@ -0,0 +1,192 @@
+//
+// This implementation is based on extendFloats() from promql/engine.go.
+func NewExtendedPointsForSmoothed(view *types.FPointRingBufferView, rangeStart, rangeEnd int64, smoothedHead, smoothedTail *promql.FPoint, memoryConsumptionTracker *limiter.MemoryConsumptionTracker) (ExtendedSmoothedPoints, error) {
+	e := ExtendedSmoothedRangeVectorPoints{
+		smoothedHead: smoothedHead,
+		smoothedTail: smoothedTail,
+	}
+	smoothedHeadSet, smoothedTailSet, err := e.extendSmoothed(view, rangeStart, rangeEnd, memoryConsumptionTracker)
+	if err != nil {
+		return ExtendedSmoothedPoints{}, err
+	}
+
+	ret := ExtendedSmoothedPoints{smoothedHeadSet: smoothedHeadSet, smoothedTailSet: smoothedTailSet}
+	ret.points = e.points.points
+
+	return ret, nil
+}
+
+// extend  is an internal function and should not be called from outside this object
+func (e *ExtendedRangeVectorPoints) extend(view *types.FPointRingBufferView, rangeStart, rangeEnd int64, memoryConsumptionTracker *limiter.MemoryConsumptionTracker) error {
+	head, tail := view.UnsafePoints()
+	count := len(head) + len(tail)
+
+	// We need a new buffer to store the extended points
+	// The caller is responsible for releasing this slice back to the slices pool
+	buff, err := types.FPointSlicePool.Get(count+2, memoryConsumptionTracker)
+	if err != nil {
+		return err
+	}
+
+	e.first = head[0]
+
+	// Add synthetic or clamp start boundary
+	if e.first.T == rangeStart {
+		buff = append(buff, e.first)
+	} else if e.first.T > rangeStart {
+		buff = append(buff, promql.FPoint{T: rangeStart, F: e.first.F})
+		buff = append(buff, e.first)
+	} else {
+		buff = append(buff, promql.FPoint{T: rangeStart, F: e.first.F})
+	}
+
+	buff = append(buff, head[1:]...)
+	buff = append(buff, tail...)
+
+	// Add synthetic or clamp end boundary
+	lastIdx := len(buff) - 1
+	e.last = buff[lastIdx]
+	if e.last.T < rangeEnd {
+		buff = append(buff, promql.FPoint{T: rangeEnd, F: e.last.F})
+	} else if e.last.T > rangeEnd {
+		buff[lastIdx].T = rangeEnd
+	}
+
+	e.points = ExtendedPoints{points: buff}
+	return nil
+}
+
+// extendSmoothed is an internal function and should not be called from outside this object
+func (e *ExtendedSmoothedRangeVectorPoints) extendSmoothed(view *types.FPointRingBufferView, rangeStart, rangeEnd int64, memoryConsumptionTracker *limiter.MemoryConsumptionTracker) (bool, bool, error) {
+	if err := e.extend(view, rangeStart, rangeEnd, memoryConsumptionTracker); err != nil {
+		return false, false, err
+	}
+
+	// Smoothing has 2 special cases.
+	// Firstly, the values on the boundaries are replaced with an interpolated values - there by smoothing the value to reflect the time of the point before/after the boundary
+	// Secondly, if vector will be used in a rate/increase function then the boundary points must be calculated differently to consider the value as a counter.
+	// These alternate points will be stored alongside the resulting vector so that the rate/increase function handler can utilise these values.
```

Comment:
These values are always calculated, right? (ie. they're not just calculated if `rate` or `increase` wrap this selector)

The way this is phrased suggests the behaviour changes if `rate` or `increase` wrap this selector, but it's the same either way.
```suggestion
	// Secondly, to ensure rate/increase return the correct values, we need to calculate and store alternate points in addition to the interpolated points.
	// These alternate points ensure that rate/increase don't incorrectly detect counter resets at the beginning and end of the range, and will be stored alongside the resulting vector so that the rate/increase function handler can utilise these values.
	// This is done regardless of whether or not this selector is wrapped in rate/increase.
```